File: src/main/java/com/example/legacyshop/service/DiscountService.java
package com.example.legacyshop.service;

import com.example.legacyshop.config.BusinessConfig;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Service for calculating discounts based on config-driven business rules.
 * 
 * Key patterns:
 * - Config-driven business logic
 * - Tiered discount calculation
 * - Immutable discount calculation
 */
@Service
public class DiscountService {

    private final BusinessConfig businessConfig;

    public DiscountService(BusinessConfig businessConfig) {
        this.businessConfig = businessConfig;
    }

    /**
     * Calculate discount amount based on order subtotal.
     * Uses tiered discount structure from configuration.
     */
    public BigDecimal calculateDiscount(BigDecimal subtotal) {
        if (subtotal == null) {
            return BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP);
        }

        BusinessConfig.Promotions promotions = businessConfig.getPromotions();
        
        // Check highest tier first (tier3)
        if (subtotal.compareTo(promotions.getTier3().getThreshold()) >= 0) {
            return subtotal.multiply(promotions.getTier3().getDiscount())
                    .setScale(2, RoundingMode.HALF_UP);
        }
        
        // Check middle tier (tier2)
        if (subtotal.compareTo(promotions.getTier2().getThreshold()) >= 0) {
            return subtotal.multiply(promotions.getTier2().getDiscount())
                    .setScale(2, RoundingMode.HALF_UP);
        }
        
        // Check lowest tier (tier1)
        if (subtotal.compareTo(promotions.getTier1().getThreshold()) >= 0) {
            return subtotal.multiply(promotions.getTier1().getDiscount())
                    .setScale(2, RoundingMode.HALF_UP);
        }
        
        // No discount applicable
        return BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * Get the discount rate (percentage) that would apply to the given subtotal.
     * Useful for displaying to customers.
     */
    public BigDecimal getDiscountRate(BigDecimal subtotal) {
        if (subtotal == null) {
            return BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP);
        }

        BusinessConfig.Promotions promotions = businessConfig.getPromotions();
        
        if (subtotal.compareTo(promotions.getTier3().getThreshold()) >= 0) {
            return promotions.getTier3().getDiscount().setScale(2, RoundingMode.HALF_UP);
        }
        
        if (subtotal.compareTo(promotions.getTier2().getThreshold()) >= 0) {
            return promotions.getTier2().getDiscount().setScale(2, RoundingMode.HALF_UP);
        }
        
        if (subtotal.compareTo(promotions.getTier1().getThreshold()) >= 0) {
            return promotions.getTier1().getDiscount().setScale(2, RoundingMode.HALF_UP);
        }
        
        return BigDecimal.ZERO.setScale(2, RoundingMode.HALF_UP);
    }
}

File: legacyshop-api/app/services/discount_service.py
"""
Discount calculation service.
Implements tiered discount logic matching the Java implementation.
"""

from decimal import Decimal, ROUND_HALF_UP
from app.config.business_config import config


class DiscountService:
    """Service for calculating discounts based on order subtotal"""
    
    def calculate_discount(self, subtotal: Decimal) -> Decimal:
        """
        Calculate discount amount based on order subtotal.
        Uses tiered discount structure from configuration.
        
        Returns the discount amount (not the rate).
        """
        if subtotal is None:
            return Decimal("0.00").quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
        
        if subtotal >= config.tier3_threshold:
            discount = (subtotal * config.tier3_discount).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
            return discount
        
        if subtotal >= config.tier2_threshold:
            discount = (subtotal * config.tier2_discount).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
            return discount
        
        if subtotal >= config.tier1_threshold:
            discount = (subtotal * config.tier1_discount).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
            return discount
        
        return Decimal("0.00").quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)


discount_service = DiscountService()
